### DISCUSSION TITLE
STM32F411CEU6 how wakeup deepsleep by PA0?

### DISCUSSION THREAD

**Discussion by xtbwqtq** [micropython/micropython#15524] [STM32 / Pyboard] [answered]
```
import stm

A0 = Pin('A0', Pin.IN, pull=Pin.PULL_DOWN) 
stm.mem32[stm.PWR+0x08] |= 1
```
this code is not work

**Comment by davefes** [comment]
I was given this on the forum:
```
You can exit from normal machine.sleep() using any pin, and either rising or falling, like this:

p = machine.Pin('X1', machine.Pin.IN, machine.Pin.PULL_UP)
p.irq(trigger=machine.Pin.IRQ_RISING, handler=lambda p:None)
machine.sleep()

But to exit from machine.deepsleep() you must use X1 (=PA0) and it requires special code to enable this feature:

import stm
stm.mem32[stm.PWR + stm.PWR_CSR] |= 1 << 8 # enable WKUP pin on PA0
machine.deepsleep()

X1 will be in input pull-down mode and will wake the MCU when it goes high. Unfortunately there is no way to make it pull-up and wake when it goes low (deep sleep is very low power consumption and only a handful of features are enabled during this mode).
```
Is `[stm.PWR+0x08]` the same as `stm.PWR_CSR`?

**Reply by davefes** [reply]
Are you using:
https://github.com/nspsck/STM32F411CEU6_BlackPill_Micropython

**Reply by xtbwqtq** [reply]
```
from pyb import LED
import time,machine

import stm
stm.mem32[stm.PWR + stm.PWR_CSR] |= 0x100

LED(1).on()
time.sleep(10)


machine.deepsleep()
```
this code still no wakeup ...

**Reply by xtbwqtq** [reply]
> Are you using: https://github.com/nspsck/STM32F411CEU6_BlackPill_Micropython

Very close to this，:   https://github.com/WeActStudio/WeActStudio.MiniSTM32F4x1

**Reply by xtbwqtq** [reply]
![image](https://github.com/user-attachments/assets/b23df302-355a-4ef1-8c7b-33a7414f6f46)

**Reply by davefes** [reply]
Found where I got the information:
https://github.com/micropython/micropython/issues/3834
There seems to be more to it than what I quoted.

**Reply by xtbwqtq** [reply]
I've seen this discussion, but it doesn't solve my problem.

It uses

```
stm.mem32[stm.PWR + stm.PWR_CSR] |= 1 << 8
```
but it not work ，in my board....

**Reply by xtbwqtq** [reply]
rtc.wakeup(60000)  is be worked,but my condition requires PA0

**Reply by davefes** [reply, marked as answer]
> X1 will be in input pull-down mode and will wake the MCU when it goes high. Unfortunately there is no way to make it pull-up and wake when it goes low (deep sleep is very low power consumption and only a handful of features are enabled during this mode).

Your circuit shows User KEY pulling PA0 low.

**Reply by xtbwqtq** [reply]
I will try to use an external button instead of the key on the board.

**Reply by davefes** [reply]
I would put the switch in series with a 10K resistor up to 3V3.  Sorry, it has been 6 months since ! was looking at this issue.

**Reply by xtbwqtq** [reply]
I used an external button, connected to the PA0 pin, and it worked fine!

thanks !!!!!

**Reply by davefes** [reply]
Great.  I also noticed that in `boot.py` I have the following:
```py
# This file is executed on every boot (including wake-boot from deepsleep)

#import esp
#esp.osdebug(None)

import pyb
pyb.usb_mode('VCP')
```

**Reply by xtbwqtq** [reply]
```
from pyb import LED,RTC,standby
import time,machine,stm,log

stm.mem32[stm.PWR + stm.PWR_CSR] |= 1 << 8

log.log(str(machine.reset_cause()))
LED(1).on()
time.sleep(10)
rtc = RTC()
rtc.wakeup(10 * 1000, None)

standby()
```
But now there is another problem. When I use ```rtc.wakeup``` at the same time, it will restart continuously, as if it is not compatible with ```stm.mem32[stm.PWR + stm.PWR_CSR] |= 1 << 8```。。

**Reply by davefes** [reply]
I never tried doing that.  You should probably start a new thread and see if an expert can help.
